File: samples/sandbox/app/src/main/java/com/scichart/docsandbox/examples/java/series3d/WaterfallSeries3D.java
package com.scichart.docsandbox.examples.java.series3d;

import androidx.annotation.NonNull;

import com.scichart.charting3d.model.dataSeries.waterfall.WaterfallDataSeries3D;
import com.scichart.charting3d.visuals.SciChartSurface3D;
import com.scichart.charting3d.visuals.pointMarkers.SpherePointMarker3D;
import com.scichart.charting3d.visuals.renderableSeries.data.GradientColorPalette;
import com.scichart.charting3d.visuals.renderableSeries.data.SolidColorBrushPalette;
import com.scichart.charting3d.visuals.renderableSeries.waterfall.WaterfallRenderableSeries3D;
import com.scichart.docsandbox.core.ExampleDefinition;
import com.scichart.docsandbox.examples.base.SingleChart3DFragment;

@ExampleDefinition()
public class WaterfallSeries3D extends SingleChart3DFragment {
    private static final int POINTS_PER_SLICE = 128;
    private static final int SLICE_COUNT = 20;

    @Override
    protected void initExample(@NonNull SciChartSurface3D surface) {
        createSurfaceMeshSeries3D();
    }

    void createSurfaceMeshSeries3D() {
        // <CreateSurfaceMeshSeries3D>
        final GradientColorPalette fillColorPalette = new GradientColorPalette(new int[]{0xFFFF0000, 0xFFFFA500, 0xFFFFFF00, 0xFFADFF2F, 0xFF006400}, new float[]{0.0f, 0.25f, 0.5f, 0.75f, 1.0f});
        final GradientColorPalette strokeColorPalette = new GradientColorPalette(new int[]{0xFFDC143C, 0xFFFF8C00, 0xFF32CD32, 0xFF32CD32}, new float[]{0.0f, 0.33f, 0.67f, 1.0f});

        final WaterfallDataSeries3D<Double, Double, Double> ds = new WaterfallDataSeries3D<>(Double.class, Double.class, Double.class, POINTS_PER_SLICE, SLICE_COUNT);
        ds.setStartX(10.0);
        ds.setStartZ(1.0);

        fillDataSeries(ds);

        final WaterfallRenderableSeries3D rSeries = new WaterfallRenderableSeries3D();
        rSeries.setDataSeries(ds);
        rSeries.setStroke(0xFF0000FF);
        rSeries.setStrokeThickness(convertValueToDp(1.0f));
        rSeries.setSliceThickness(0.0f);
        rSeries.setYColorMapping(fillColorPalette);
        rSeries.setYStrokeColorMapping(strokeColorPalette);
        rSeries.setOpacity(0.8f);
        // </CreateSurfaceMeshSeries3D>

        // <ApplyingSolidPalettes>
        rSeries.setYColorMapping(new SolidColorBrushPalette(0xFF6495ED));
        rSeries.setYStrokeColorMapping(new SolidColorBrushPalette(0xFF6495ED));
        // </ApplyingSolidPalettes>

        // <ApplyingLinearGradientPalettes>
        final int[] colors = new int[]{0xFFFF0000, 0xFFFFA500, 0xFFFFFF00, 0xFFADFF2F, 0xFF006400};
        final float[] stops = new float[]{0, .25f, .5f, .75f, 0};
        final GradientColorPalette colorPalette = new GradientColorPalette(colors, stops);
        // </ApplyingLinearGradientPalettes>

        // <ApplyingColorPaletteOntoSliceFill>
        // Z-Direction
        rSeries.setZColorMapping(colorPalette);
        rSeries.setZStrokeColorMapping(new SolidColorBrushPalette(0x00FFFFFF));

        // or Y-Direction
        rSeries.setYColorMapping(colorPalette);
        rSeries.setYStrokeColorMapping(new SolidColorBrushPalette(0x00FFFFFF));
        // </ApplyingColorPaletteOntoSliceFill>

        // <ApplyingColorPaletteOntoSliceStroke>
        // Z-Direction
        rSeries.setZColorMapping(new SolidColorBrushPalette(0x00FFFFFF));
        rSeries.setZStrokeColorMapping(colorPalette);

        // or Y-Direction
        rSeries.setYColorMapping(new SolidColorBrushPalette(0x00FFFFFF));
        rSeries.setYStrokeColorMapping(colorPalette);
        // </ApplyingColorPaletteOntoSliceStroke>
    }

    private void fillDataSeries(WaterfallDataSeries3D<Double, Double, Double> ds) { }

    void applySliceThickness() {
        // <ApplySliceThickness>
        final WaterfallRenderableSeries3D rSeries = new WaterfallRenderableSeries3D();
        rSeries.setSliceThickness(10.0f);
        // </ApplySliceThickness>
    }

    void pointMarkersOnWaterfall3D() {
        // <PointMarkersOnWaterfall3D>
        final SpherePointMarker3D pointMarker = new SpherePointMarker3D();
        pointMarker.setFill(0xFFFFA500);
        pointMarker.setSize(convertValueToDp(10.0f));

        final WaterfallRenderableSeries3D rSeries = new WaterfallRenderableSeries3D();
        rSeries.setPointMarker(pointMarker);
        // </PointMarkersOnWaterfall3D>
    }
}

File: samples/sandbox/app/src/main/java/com/scichart/docsandbox/examples/kotlin/series3d/WaterfallSeries3D.kt
package com.scichart.docsandbox.examples.kotlin.series3d

import com.scichart.charting3d.model.dataSeries.waterfall.WaterfallDataSeries3D
import com.scichart.charting3d.visuals.SciChartSurface3D
import com.scichart.charting3d.visuals.pointMarkers.SpherePointMarker3D
import com.scichart.charting3d.visuals.renderableSeries.data.GradientColorPalette
import com.scichart.charting3d.visuals.renderableSeries.data.SolidColorBrushPalette
import com.scichart.charting3d.visuals.renderableSeries.waterfall.WaterfallRenderableSeries3D
import com.scichart.docsandbox.core.ExampleDefinition
import com.scichart.docsandbox.examples.base.SingleChart3DFragment

@ExampleDefinition()
class WaterfallSeries3D : SingleChart3DFragment() {
    private val POINTS_PER_SLICE = 128
    private val SLICE_COUNT = 20

    override fun initExample(surface: SciChartSurface3D) {
        createSurfaceMeshSeries3D()
    }

    fun createSurfaceMeshSeries3D() {
        // <CreateSurfaceMeshSeries3D>
        val fillColorPalette = GradientColorPalette(
            intArrayOf(0xFFFF0000.toInt(), 0xFFFFA500.toInt(), 0xFFFFFF00.toInt(), 0xFFADFF2F.toInt(), 0xFF006400.toInt()),
            floatArrayOf(0.0f, 0.25f, 0.5f, 0.75f, 1.0f)
        )
        val strokeColorPalette = GradientColorPalette(
            intArrayOf(0xFFDC143C.toInt(), 0xFFFF8C00.toInt(), 0xFF32CD32.toInt(), 0xFF32CD32.toInt()),
            floatArrayOf(0.0f, 0.33f, 0.67f, 1.0f)
        )
        val ds = WaterfallDataSeries3D(
            Double::class.java,
            Double::class.java,
            Double::class.java, POINTS_PER_SLICE, SLICE_COUNT
        )
        ds.setStartX(10.0)
        ds.setStartZ(1.0)

        fillDataSeries(ds)

        val rSeries = WaterfallRenderableSeries3D()
        rSeries.dataSeries = ds
        rSeries.stroke = -0xffff01
        rSeries.strokeThickness = convertValueToDp(1.0f)
        rSeries.sliceThickness = 0.0f
        rSeries.yColorMapping = fillColorPalette
        rSeries.yStrokeColorMapping = strokeColorPalette
        rSeries.opacity = 0.8f
        // </CreateSurfaceMeshSeries3D>

        // <ApplyingSolidPalettes>
        rSeries.yColorMapping = SolidColorBrushPalette(0xFF6495ED.toInt())
        rSeries.yStrokeColorMapping = SolidColorBrushPalette(0xFF6495ED.toInt())
        // </ApplyingSolidPalettes>

        // <ApplyingLinearGradientPalettes>
        val colors = intArrayOf(0xFFFF0000.toInt(), 0xFFFFA500.toInt(), 0xFFFFFF00.toInt(), 0xFFADFF2F.toInt(), 0xFF006400.toInt())
        val stops = floatArrayOf(0f, .25f, .5f, .75f, 0f)
        val colorPalette = GradientColorPalette(colors, stops)
        // </ApplyingLinearGradientPalettes>

        // <ApplyingColorPaletteOntoSliceFill>
        // Z-Direction
        rSeries.zColorMapping = colorPalette
        rSeries.zStrokeColorMapping = SolidColorBrushPalette(0x00FFFFFF)

        // or Y-Direction
        rSeries.yColorMapping = colorPalette
        rSeries.yStrokeColorMapping = SolidColorBrushPalette(0x00FFFFFF)
        // </ApplyingColorPaletteOntoSliceFill>

        // <ApplyingColorPaletteOntoSliceStroke>
        // Z-Direction
        rSeries.zColorMapping = SolidColorBrushPalette(0x00FFFFFF)
        rSeries.zStrokeColorMapping = colorPalette

        // or Y-Direction
        rSeries.yColorMapping = SolidColorBrushPalette(0x00FFFFFF)
        rSeries.yStrokeColorMapping = colorPalette
        // </ApplyingColorPaletteOntoSliceStroke>
    }

    private fun fillDataSeries(ds: WaterfallDataSeries3D<Double, Double, Double>) {}

    fun applySliceThickness() {
        // <ApplySliceThickness>
        val rSeries = WaterfallRenderableSeries3D()
        rSeries.sliceThickness = 10.0f
        // </ApplySliceThickness>
    }

    fun pointMarkersOnWaterfall3D() {
        // <PointMarkersOnWaterfall3D>
        val pointMarker = SpherePointMarker3D()
        pointMarker.fill = 0xFFFFA500.toInt()
        pointMarker.size = convertValueToDp(10.0f)

        val rSeries = WaterfallRenderableSeries3D()
        rSeries.pointMarker = pointMarker
        // </PointMarkersOnWaterfall3D>
    }
}
